packages/dd-trace/src/llmobs/span_processor.js

```diff
@@ -0,0 +1,147 @@
+'use strict'
+
+const {
+  SPAN_KIND,
+  MODEL_NAME,
+  MODEL_PROVIDER,
+  METADATA,
+  INPUT_MESSAGES,
+  INPUT_VALUE,
+  OUTPUT_MESSAGES,
+  INPUT_DOCUMENTS,
+  OUTPUT_DOCUMENTS,
+  OUTPUT_VALUE,
+  METRICS,
+  ML_APP,
+  TAGS,
+  PARENT_ID_KEY,
+  SESSION_ID,
+  NAME
+} = require('./constants')
+
+const {
+  ERROR_MESSAGE,
+  ERROR_TYPE,
+  ERROR_STACK
+} = require('../constants')
+
+const AgentlessWriter = require('./writers/spans/agentless')
+const AgentProxyWriter = require('./writers/spans/agentProxy')
+const { isLLMSpan } = require('./util')
+
+const tracerVersion = require('../../../../package.json').version
+
+class LLMObsSpanProcessor {
+  constructor (config) {
+    this._config = config
+    const { llmobs } = config
+
+    if (llmobs.enabled) {
+      const LLMObsSpanWriter = llmobs.agentlessEnabled ? AgentlessWriter : AgentProxyWriter
+      this._writer = new LLMObsSpanWriter(config)
+    }
+  }
+
+  process (span) {
+    if (!this._config.llmobs.enabled) return
+    if (!isLLMSpan(span)) return
+    const payload = this._process(span)
+
+    this._writer.append(payload)
+  }
+
+  _process (span) {
+    const tags = span.context()._tags
+    const spanKind = tags[SPAN_KIND]
+
+    const meta = { 'span.kind': spanKind, input: {}, output: {} }
+    const input = {}
+    const output = {}
+
+    if (['llm', 'embedding'].includes(spanKind) && tags[MODEL_NAME]) {
+      meta.model_name = tags[MODEL_NAME]
+      meta.model_provider = (tags[MODEL_PROVIDER] || 'custom').toLowerCase()
+    }
+    if (tags[METADATA]) {
+      meta.metadata = JSON.parse(tags[METADATA])
```

Comment:
[APM input needed] Coming back to the whole `JSON.stringify` and then now `JSON.parse`... I think on the tagger I stringified because tags _should_ be set as strings. However, if I won't be adding these LLMObs tags to the APM payload anyways, can I just store these JSON properties as objects on the span tags? Then, I won't have to do the whole stringify/parse game, and probably boost performance a little.

Reply:
Roch and I talked offline - we'll try and do this in follow-up work next week. For now, we'll store these on tags.